src/Data/List.purs

```diff
@@ -632,18 +637,65 @@ tails list@(Cons _ tl)= list : tails tl
 --------------------------------------------------------------------------------
 
 -- | Remove duplicate elements from a list.
+-- | Keeps the first duplicates found and preserves ordering otherwise.
+-- |
+-- | ```purescript
+-- | nub 1:2:1:3:3:Nil = 1:2:3:Nil
+-- | ```
+-- |
+-- | Running time: `O(n log n)`
+nub :: forall a. Ord a => List a -> List a
+nub = nubBy compare
+
+-- | Remove duplicate elements from a list based on the provided ordering function.
+-- | Keeps the first duplicates found and preserves ordering otherwise.
+-- |
+-- | ```purescript
+-- | nubBy (compare `on` Array.length) ([1]:[2]:[3,4]:Nil) == [1]:[3,4]:Nil
+-- | ```
+-- |
+-- | Running time: `O(n log n)`
+nubBy :: forall a. (a -> a -> Ordering) -> List a -> List a
+nubBy p =
+  -- Add indices so we can recover original order after deduplicating.
+  addIndexReverse
+  -- Sort by original values to cluster duplicates.
+  >>> sortBy (p `on` snd)
+  -- Removing neighboring duplicates.
+  >>> nubByAdjacentReverse (\a b -> (p `on` snd) a b == EQ)
+  -- Sort by index to recover original order.
+  -- Use `flip` to sort in reverse order in anticipation of final `mapReverse`.
+  >>> sortBy (flip compare `on` fst)
+  -- Discard indicies, just keep original values.
+  >>> mapReverse snd
```

Comment:
I can't think of a much faster way to write this.
We could instead just use the more naive versions of the `O(n)` functions, considering that the sorts (`O(n log n)`) will take up the bulk of the time. But I figured it would be best to try for efficiency when possible.